### PR TITLE
Added more methods for Handler and added support for embeds and components

### PR DESCRIPTION
**Pull request by Rashaad1268**

### CODE REVIEW

File: discord/embeds.py [hunk 1]
```diff
@@ -72,10 +73,21 @@ def set_thumbnail(self, *, url, height: int = None, width: int = None):
...
         for field in self.fields:
             length += len(field["name"]) + len(field["value"])
-        length += len(self.footer["text"]) + len(self.author["name"]) + len(self.timestamp)
+
+        if hasattr(self, "footer"):
+            length += len(self.footer["text"])
+        if hasattr(self, "author"):
+            length += len(self.author["name"])
+        if hasattr(self, "timestamp"):
+            length += len(self.timestamp) if self.timestamp else 0
```
Comment (sebkuip) [inline discord/embeds.py:90]: Why is there the extra `if self.timestamp else 0` when you already checked with hasattr?

File: discord/handler.py [hunk 2]
```diff
@@ -65,13 +78,78 @@ async def send_message(self, channel_id: int, content: typing.Optional[str] = No
         except KeyError:
             return data
 
+    async def edit_message(self, channel_id: int, message_id: int, content: Optional[str] = None,
```
Comment (sebkuip) [inline discord/handler.py:81]: Could the optional parameters be made kwargs?

**Review by sebkuip — CHANGES_REQUESTED** [review]

File: discord/embeds.py (hunk 1 above)
Comment (Rashaad1268) [reply discord/embeds.py:90]: `self.timestamp` can be `None`

Comment (Rashaad1268) [reply discord/embeds.py:90]: yeah the `hasattr(self, "timestamp")` is not needed

**Review by sebkuip — APPROVED** [review]

**Comment by sebkuip** [comment]
I will merge this once the big PR for a lot of models is in too. Hang tight!

File: discord/components.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
+import os
+
+
+class Component:
```
Comment (sebkuip) [inline discord/components.py:4]: This file doesn't fit in the planned file setup. Please look at #7 for how we are going to structure the models

File: discord/handler.py [hunk 4]
```diff
@@ -3,14 +3,15 @@
...
+from .components import View
 
 
 class Handler:
     def __init__(self):
         self.base_url: str = 'https://discord.com/api/v9/'
         self.user_agent: str = "Disthon test library V0.0.1b"
 
-    async def request(self, method: str, dest: str, *, headers: typing.Optional[dict] = None,
+    async def request(self, method: str, dest: str, *, headers: Optional[dict] = None,
```
Comment (sebkuip) [inline discord/handler.py:14]: Either remove all typing, or keep the typing here

File: discord/handler.py [hunk 5]
```diff
@@ -55,19 +56,26 @@ async def connect(self, url: str) -> aiohttp.ClientWebSocketResponse:
     async def close(self) -> None:
         await self.__session.close()
 
-    async def send_message(self, channel_id: int, content: typing.Optional[str] = None, embed: Optional[Embed] = None,
-                           embeds: Optional[typing.List[Embed]] = None):
+    async def send_message(self, channel_id: int, content: Optional[str] = None, embed: Optional[Embed] = None,
```
Comment (sebkuip) [inline discord/handler.py:59]: Follow up on previous point on typing

File: discord/handler.py [hunk 6]
```diff
@@ -55,19 +56,26 @@ async def connect(self, url: str) -> aiohttp.ClientWebSocketResponse:
...
+                           embeds: Optional[List[Embed]] = None, view: Optional[View] = None,
+                           views: Optional[List[View]] = None):
         if embeds is None:
             embeds = []
         if embed:
             embeds.append(embed)
+        if views is None:
+            views = []
+        if view:
+            views.append(view)
```
Comment (sebkuip) [inline discord/handler.py:69]: views is not defined

File: discord/handler.py [hunk 7]
```diff
@@ -55,19 +56,26 @@ async def connect(self, url: str) -> aiohttp.ClientWebSocketResponse:
...
 
         payload = {}
 
         if content:
             payload["content"] = content
         if embeds:
             payload["embeds"] = list(map(lambda e: e._to_dict(), embeds))
+        if views:
+            payload["components"] = [view._to_dict() for view in views]
```
Comment (sebkuip) [inline discord/handler.py:78]: a list(map()) would be more in-line with the rest of the code

File: discord/handler.py [hunk 8]
```diff
@@ -79,9 +87,14 @@ async def send_message(self, channel_id: int, content: typing.Optional[str] = No
             return data
 
     async def edit_message(self, channel_id: int, message_id: int, *, content: Optional[str] = None,
-                           embed: Optional[Embed] = None, embeds: Optional[List[Embed]] = None):
+                           embed: Optional[Embed] = None, embeds: Optional[List[Embed]] = None,
```
Comment (sebkuip) [inline discord/handler.py:90]: Same point as before on typing

**Review by sebkuip — CHANGES_REQUESTED** [review]
Most seems fine to me, except that `disthon/discord/interactions/__init__.py` is empty. I would at least import the view from there. So when people do `from discord import interactions` it will import the correct classes, and people can make a View more easily.

**Comment by AA1999** [comment]
Why a embed and a embeds at the same time?

File: discord/handler.py (hunk 4 above)
Comment (AA1999) [reply discord/handler.py:13]: I prefer if he did keep all the typing

**Review by AA1999 — CHANGES_REQUESTED** [review]
Please check the requested changes for each method/class

File: discord/handler.py [hunk 9]
```diff
@@ -51,12 +55,28 @@ async def connect(self, url: str) -> aiohttp.ClientWebSocketResponse:
     async def close(self) -> None:
         await self.__session.close()
 
-    async def send_message(self, channel_id: int, content: typing.Optional[str] = None):
-        payload = {
-            'content': content,
-        }
+    async def send_message(self, channel_id: int, content: Optional[str] = None, embed: Optional[Embed] = None,
```
[the remark this answers is not in the document]
Comment (Rashaad1268) [reply discord/handler.py:58]: It will be easier if the user only needs to send one embed

[the remark this answers is not in the document]
Comment (AA1999) [reply discord/handler.py:58]: Maybe make it a Union of `Embed, list[Embed]`

File: discord/embeds.py [hunk 10]
```diff
@@ -72,13 +73,26 @@ def set_thumbnail(self, *, url, height: int = None, width: int = None):
...
+        if hasattr(self, "footer"):
+            length += len(self.footer["text"])
+        if hasattr(self, "author"):
+            length += len(self.author["name"])
         return length
 
-    def from_dict(self, data: Dict[str, Any]):
+    @classmethod
+    def from_dict(cls, data: Dict[str, Any]):
+        self = cls()
```
[the remark this answers is not in the document]
Comment (AA1999) [reply discord/embeds.py:95]: Working on Embeds myself, close to what dpy did tbh
But feel free to continue working on it so we can merge ideas

**Review by AA1999 — CHANGES_REQUESTED** [review]
Please check the previous reviews and fix the code, it's still not fixed

File: discord/embeds.py [hunk 11]
```diff
@@ -15,13 +15,13 @@ def __init__(
...
-        self.type = type
-        self.url = url
-        self.fields = []
+        self.title: Optional[str] = str(title) if title else None
+        self.description: Optional[str] = str(description) if description else None
+        self.timestamp: Optional[str] = str(timestamp) if timestamp else None
+        self.color: Optional[int] = color
+        self.type: str = type
+        self.url: Optional[str] = url
+        self.fields: list = []
```
Comment (AA1999) [inline discord/embeds.py:24]: There are a few more fields for the embed that I included in the new file, try improving on that

File: discord/interactions/components.py [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+        self.disabled = disabled
+        self.style = style
+        self.label = label
+        self.emoji = emoji
+        self.url = url
+        self.options = options
+        self.placeholder = placeholder
+        self.min_values = min_values
+        self.max_values = max_values
+        self.custom_id = custom_id
```
Comment (AA1999) [inline discord/interactions/components.py:23]: Typehint disabled to custom_id please

File: discord/embeds.py [hunk 13]
```diff
@@ -1,92 +1,305 @@
...
+        if name is None:
+            return name
+        if isinstance(name, str):
+            if len(name) == 0:
+                raise EmptyField
+            else:
+                return name
+        raise ValueError
+
+    @validator('url')
```
Comment (AA1999) [inline discord/embeds.py:96]: Same as above validator

File: discord/embeds.py [hunk 14]
```diff
@@ -1,92 +1,305 @@
...
+        if len(match) != 1:
+            raise InvalidURL
+        return url
+
+
+class EmbedProvider(BaseModel):
+    name: Optional[str] = None
+    url: Optional[str] = None
+
+    @validator('url')
```
Comment (AA1999) [inline discord/embeds.py:68]: This should be converted into a @classmethod

**Review by AA1999 — CHANGES_REQUESTED** [review]
Needs changes to the validators and typehints

**Review by sebkuip — APPROVED** [review]

**Comment by sebkuip** [comment]
Merged into import-fix. Closing PR. Thanks!

**Pull request closed without merging**